test/test_connection.py

```diff
@@ -0,0 +1,242 @@
+# -*- coding: utf-8 -*-
+
+import pytest
+
+from wsproto.connection import WSConnection, CLIENT, SERVER, ConnectionState
+from wsproto.events import (ConnectionClosed, TextReceived, BytesReceived)
+from wsproto.frame_protocol import CloseReason, FrameProtocol
+
+
+class FakeProtocol(object):
```

Comment:
Interesting... in `h11` I wrote these kinds of tests by having a bit of utility code to glue together two connection objects and do the initial setup, and then I would like, inject weird data or make assertions about the data on the wire. ([Utility code](https://github.com/njsmith/h11/blob/05ad7a72a30e4fc3db735d85fc3553bd5d10b164/h11/tests/helpers.py#L39), [example test](https://github.com/njsmith/h11/blob/05ad7a72a30e4fc3db735d85fc3553bd5d10b164/h11/tests/test_connection.py#L75))

This potentially has some advantages, since it makes fewer assumptions about the internals of the connection object, and thus hopefully survives better as the code evolves. Not every test can be written this way, but it might be a useful tool to have around.

I'm not suggesting you need to rewrite all these or anything like that, just bringing it up as something for us to think about :-)